### PR TITLE
Add storage recommendations and applications

### PR DESCRIPTION
**Pull request by lpsantil**
This is based, in large part, on information found here [0].

[0] https://docs.google.com/spreadsheets/d/1pTAro4ONaCCVBW3f9B9SxDT95F4dcMP3dJ_D_v_QPjo/edit#gid=0

### CODE REVIEW

**Comment by lpsantil** [comment]
@sabre1041 @etsauer @cpitman @sferich888 @raffaelespazzoli @juanluisvaladas @dstockdreher

File: scaling_performance/optimizing_storage.adoc [hunk 1]
```diff
@@ -13,6 +13,81 @@ toc::[]
...
+WARNING: ^(3)^ Vendor GlusterFS, Vendor S3, and Vendor Swift supportability and configurability may vary.
+
+=== Storage Backend Recommendations
+Below is a matrix that summarizes which storage backend types are Recommended, Supported, and Configurable for the given
+{product-title} application and useful attributes.
+[format="csv",width="70%",cols="10"]
+[options="header]
+|=======================
+Type,Tied to Node?,ROM,RWX,Registry,Scaled Registry,etcd,Metrics,Logging,Apps
+hostPath,Yes,No,No,Not Configurable,Not Configurable,Recommended,Recommended,Recommended,Configurable
```
Comment (juanluisvaladas) [inline scaling_performance/optimizing_storage.adoc]: Hostpath is actually configurable for the registry for a single instance. I'd say instead not recommended or not HA.

File: scaling_performance/optimizing_storage.adoc [hunk 2]
```diff
@@ -13,6 +13,81 @@ toc::[]
...
+== General Storage Guidance
+
+The following is a matrix which details the storage backend types {product-title} can be used with implementation examples and vendors.
+
+=== Available Storage Options
+[format="csv",width="70%",cols="3"]
+[options="header]
+|=======================
+Type,Examples,Vendors
+hostPath,"hostPath pod mounted volumes.  A specific type of Block Storage.  File paths on the node mounted onto running pods.",Others
```
Comment (juanluisvaladas) [inline scaling_performance/optimizing_storage.adoc]: I'm not a native speaker but here I understand you assign to the pod a block device instead of passing it a path that is then bound to the pod.

File: scaling_performance/optimizing_storage.adoc [hunk 3]
```diff
@@ -13,6 +13,81 @@ toc::[]
...
+===== etcd
+For etcd, the preferred Storage Backend is Block Storage.  For the best etcd reliability, the lowest consistent latency Storage Backend is preferrable.
+
+===== Metrics
+In {product-title} Hosted Metrics cluster deployment, the preferred Storage Backend is Block Storage.  All Network Attached File
+Storage Backends are expressly *not recommended* for Hosted Metrics cluster deployment.
+
+===== Logging
+In {product-title} Hosted Logging cluster deployment, the preferred Storage Backend is Block Storage.  All Network Attached File
+Storage Backends are expressly *not recommended* for Hosted Logging cluster deployment.
```
Comment (juanluisvaladas) [inline scaling_performance/optimizing_storage.adoc]: Would it make sense to add a reference for NFS storage causing data corruption?

**Review by juanluisvaladas — COMMENTED** [review]
Very good overall, there are a few things I think could be improved but that's definitely good enough to be merged.

File: scaling_performance/optimizing_storage.adoc (hunk 2 above)
Comment (raffaelespazzoli) [inline scaling_performance/optimizing_storage.adoc]: I conceptually disagree with putting hostpath at the same level of the other storage type.
as far as storage type goes hostpath is a block storage.
hostpath is a volume type, as a volume type I categorize it as a local storage.
the other relevant volume types in this discussion are
ephemeral storage (empty dir)  which is missing from your document and it's again block storage
and persistent volumes.

I suggest to talk about these concepts in a separate table.

File: scaling_performance/optimizing_storage.adoc [hunk 4]
```diff
@@ -13,6 +13,81 @@ toc::[]
...
+File,"Network Attached File Storage.  CNS GlusterFS ^(1)^, GlusterFS ^(1)^, RHEL NFS, NetApp NFS ^(2)^, Vendor NFS, Vendor GlusterFS ^(3)^, Loopback Image over NFS","RedHat, NetApp, AWS, GCE, Others"
+Object,"Object Storage.  Ceph Object Storage, OpenStack Swift,  AWS S3, Google Cloud Storage, Vendor S3 ^(3)^, Vendor Swift ^(3)^","RedHat, AWS, GCE, Others"
+|=======================
+NOTE: ^(1)^ Ceph RBD, OpenStack Cinder, AWS EBS, GCE Persistent Disk support Dynamic PV Provisioning natively in {product-title}.
+
+NOTE: ^(2)^ NetApp NFS supports Dynamic PV Provisioning when using the Trident Plugin
+
+WARNING: ^(3)^ Vendor GlusterFS, Vendor S3, and Vendor Swift supportability and configurability may vary.
+
+=== Storage Backend Recommendations
```
Comment (raffaelespazzoli) [inline scaling_performance/optimizing_storage.adoc]: as part of the application / storage recommendation discussion, you may want to also introduce this picture that is intended to show the possible combination of storage and compute
https://www.lucidchart.com/documents/edit/80e4f96d-2a84-4cbb-9608-19be6c0545c8

let me know if you have issues opening the link

File: scaling_performance/optimizing_storage.adoc [hunk 5]
```diff
@@ -13,6 +13,81 @@ toc::[]
...
+Below is a matrix that summarizes which storage backend types are Recommended, Supported, and Configurable for the given
+{product-title} application and useful attributes.
+[format="csv",width="70%",cols="10"]
+[options="header]
+|=======================
+Type,Tied to Node?,ROM,RWX,Registry,Scaled Registry,etcd,Metrics,Logging,Apps
```
Comment (raffaelespazzoli) [inline scaling_performance/optimizing_storage.adoc]: scaled registry on block storage? I don't think it works...

Comment (raffaelespazzoli) [inline scaling_performance/optimizing_storage.adoc]: for etcd I assume you don't mean the etcd that openshift uses, maybe we should specify that. If that is so, why is it relevant for this discussion?

Comment (raffaelespazzoli) [inline scaling_performance/optimizing_storage.adoc]: there shouldn't be more than one recommended option per column
logging and metrics should be recommended for block storage not hostpath

Comment (raffaelespazzoli) [inline scaling_performance/optimizing_storage.adoc]: I would remove the app column, because the recommendation is really app dependent

File: scaling_performance/optimizing_storage.adoc (hunk 2 above)
Comment (rootfs) [inline scaling_performance/optimizing_storage.adoc]: hostPath is not meant for production, it is more a developer's favorite.

File: scaling_performance/optimizing_storage.adoc [hunk 6]
```diff
@@ -13,6 +13,81 @@ toc::[]
...
+The following is a matrix which details the storage backend types {product-title} can be used with implementation examples and vendors.
+
+=== Available Storage Options
+[format="csv",width="70%",cols="3"]
+[options="header]
+|=======================
+Type,Examples,Vendors
+hostPath,"hostPath pod mounted volumes.  A specific type of Block Storage.  File paths on the node mounted onto running pods.",Others
+Block,"Block Storage. HDD, SSD, NVMe, Raw Physical Disk, VM Physical Disk, VMDK, LVM, iSCSI, Fibre Channel, Ceph RBD, OpenStack Cinder, Dell/EMC Scale.IO, AWS EBS ^(1)^, GCE Persistent Disk ^(1)^, Azure Disk","RedHat, Seagate, WesternDigital, Toshiba, HGST, Intel, Samsung, OCZ, Plextor, PNY, Kingston, SanDisk, AWS, GCE, VMware, Dell/EMC, HPE, Others"
+File,"Network Attached File Storage.  CNS GlusterFS ^(1)^, GlusterFS ^(1)^, RHEL NFS, NetApp NFS ^(2)^, Vendor NFS, Vendor GlusterFS ^(3)^, Loopback Image over NFS","RedHat, NetApp, AWS, GCE, Others"
```
Comment (rootfs) [inline scaling_performance/optimizing_storage.adoc]: CephFS (still tech preview from RHS), Azure File (i.e. SMB), OpenStack Manila

File: scaling_performance/optimizing_storage.adoc [hunk 7]
```diff
@@ -13,6 +13,81 @@ toc::[]
...
+Object,"Object Storage.  Ceph Object Storage, OpenStack Swift,  AWS S3, Google Cloud Storage, Vendor S3 ^(3)^, Vendor Swift ^(3)^","RedHat, AWS, GCE, Others"
+|=======================
+NOTE: ^(1)^ Ceph RBD, OpenStack Cinder, AWS EBS, GCE Persistent Disk support Dynamic PV Provisioning natively in {product-title}.
```
Comment (rootfs) [inline scaling_performance/optimizing_storage.adoc]: Also Glusterfs, Azure Disk

File: scaling_performance/optimizing_storage.adoc [hunk 8]
```diff
@@ -13,6 +13,81 @@ toc::[]
...
+
+The following is a matrix which details the storage backend types {product-title} can be used with implementation examples and vendors.
+
+=== Available Storage Options
+[format="csv",width="70%",cols="3"]
+[options="header]
+|=======================
+Type,Examples,Vendors
+hostPath,"hostPath pod mounted volumes.  A specific type of Block Storage.  File paths on the node mounted onto running pods.",Others
+Block,"Block Storage. HDD, SSD, NVMe, Raw Physical Disk, VM Physical Disk, VMDK, LVM, iSCSI, Fibre Channel, Ceph RBD, OpenStack Cinder, Dell/EMC Scale.IO, AWS EBS ^(1)^, GCE Persistent Disk ^(1)^, Azure Disk","RedHat, Seagate, WesternDigital, Toshiba, HGST, Intel, Samsung, OCZ, Plextor, PNY, Kingston, SanDisk, AWS, GCE, VMware, Dell/EMC, HPE, Others"
```
Comment (rootfs) [inline scaling_performance/optimizing_storage.adoc]: Raw devices are not supported yet

File: scaling_performance/optimizing_storage.adoc (hunk 2 above)
Comment (lpsantil) [inline scaling_performance/optimizing_storage.adoc]: @juanluisvaladas It can be one or the other.  https://kubernetes.io/docs/concepts/storage/volumes/#hostpath

@rootfs Current Hosted Metrics and Logging documentation are (intentionally?) vague here.  Some have implemented those with hostPath when faced with having a cluster that otherwise only has access to NFS storage.  And it is used at scale.  Those implementers have also created labeling and scheduler affinity/anti-affinity rules to support that use-case.  Part of why that was done (probably) was that there was no other obvious guidance.  I'm hope this PR rectifies that for some future implementers.

Comment (lpsantil) [inline scaling_performance/optimizing_storage.adoc]: @raffaelespazzoli Thanks for this.  I intended to only speak to Persistent Storage Types.  I need to speak with you (and others) more about ephemeral storage, because, while it may have benefits I assume its supportability is much different that Persistent Storage Types.  You can always add an entry to the spreadsheet I linked to as well.

File: scaling_performance/optimizing_storage.adoc (hunk 8 above)
Comment (lpsantil) [inline scaling_performance/optimizing_storage.adoc]: @rootfs I'll remove them.  However, I had a specific request to add VMDK.  Please see spreadsheet.

File: scaling_performance/optimizing_storage.adoc (hunk 6 above)
Comment (lpsantil) [inline scaling_performance/optimizing_storage.adoc]: @rootfs The guidance I received on CephFS was different.  Please see spreadsheet.   I see Azure File was added to https://docs.openshift.org/latest/architecture/additional_concepts/storage.html#types-of-persistent-volumes but OSP Manila was not.

File: scaling_performance/optimizing_storage.adoc (hunk 7 above)
Comment (lpsantil) [inline scaling_performance/optimizing_storage.adoc]: @rootfs Thanks!  Good catch.

File: scaling_performance/optimizing_storage.adoc (hunk 2 above)
Comment (lpsantil) [inline scaling_performance/optimizing_storage.adoc]: @raffaelespazzoli On hostPath, I agree it's a subtype of of Block Storage, and, like you mentioned, I wanted to call it out because it enjoy's its own Volume Type.

File: scaling_performance/optimizing_storage.adoc (hunk 5 above)
Comment (lpsantil) [inline scaling_performance/optimizing_storage.adoc]: @raffaelespazzoli ACK.

Comment (lpsantil) [inline scaling_performance/optimizing_storage.adoc]: @raffaelespazzoli I did intend to reference the OCP internal etcd. However, you are right.  I will remove etcd from this table.

Comment (lpsantil) [inline scaling_performance/optimizing_storage.adoc]: That's the goal.  However, hostPath is also a subtype of Block Storage.

Comment (lpsantil) [inline scaling_performance/optimizing_storage.adoc]: I almost agree with you.  But want OCP to run apps.  I'd think they'd want to know if ever that a certain storage type is less than recommended or has a caveat.  And there are two in the table as it stands.

File: scaling_performance/optimizing_storage.adoc (hunk 1 above)
Comment (lpsantil) [inline scaling_performance/optimizing_storage.adoc]: Noted.

File: scaling_performance/optimizing_storage.adoc (hunk 3 above)
Comment (lpsantil) [inline scaling_performance/optimizing_storage.adoc]: I'll add a note later in the text unless you think it should added somewhere here.

**Comment by childsb** [comment]
References to block storage should be changed to file-on-block since raw block storage isn't currently a feature (but will be sometime in the future).

**Comment by lpsantil** [comment]
@childsb Please, pound on this PR like a rock polisher until it's shiny.

I get that the language is not as specific as it could be and that the apps would only ever see files & file paths.  I've tried to write it at a level where a new(er) OCP administrator would be familiar with.  In a virtualized & contrainerized world, almost everything gets accessed over a network.  But if I said Block Device Storage, you'd understand I wasn't talking about NFS or SMB and you might know LUNs, iSCSI & Fibre Channel, etc.  And if you're an AWS Admin, you'd probably make the connection to EBS.  

Yeah, I get that Object Storage is something that isn't accessed over a PV and so I'm lumping storage types together.  But the Internal Docker Registry can use all of these type and have very different performance characteristics.  What's worse is that it's not even clear that you can use all of these types.  The docs currently say "attach a remote volume" [0].   None the wiser that the first type they have to choose from, NFS [1] is actually not supported.  More confusing is that the very next section in [0] shows how to define an AWS S3 bucket for use with the Internal Docker Registry even though it is not listed as the supported in [1] and it is even the preferred storage type if available.

In short, if you have a better way of organizing the story, I'm all ears and ready to write it that way.  Clearer docs is my goal and the subject is already a minefield. Please also look the internal spreadsheet I've started.


[0] https://docs.openshift.com/container-platform/3.5/install_config/registry/deploy_registry_existing_clusters.html#registry-production-use

[1] https://docs.openshift.com/container-platform/3.5/install_config/registry/deploy_registry_existing_clusters.html#registry-production-use

File: scaling_performance/optimizing_storage.adoc (hunk 4 above)
Comment (lpsantil) [inline scaling_performance/optimizing_storage.adoc]: I can open the link but I think I need your guidance in understanding some of the details.

**Comment by lpsantil** [comment]
So I found that our docs are using [0] Shared Storage and Block Storage.  Is that a bit better?  Object Storage isn't even mentioned.

[0] https://docs.openshift.com/container-platform/3.5/security/storage.html

File: scaling_performance/optimizing_storage.adoc (hunk 2 above)
Comment (cpitman) [reply scaling_performance/optimizing_storage.adoc]: How is it block storage? Does OCP modify permissions on host mount storage? If not, then it isn't block.

**Comment by tmorriso-rh** [comment]
@lpsantil Hi, I am following up on this PR. Is this information still valid? From what I can tell, the information just needs to be reviewed and edited. Please advise. Thanks.

**Comment by tmorriso-rh** [comment]
This is now a new PR: https://github.com/openshift/openshift-docs/pull/5623
